### PR TITLE
feat(lidar_centerpoint): accelerated preprocessing for centerpoint

### PR DESCRIPTION
**Pull request by knzo25**
## Description
While changing the point type to the new standard, realized that preprocessing was bottleneck in centerpoint.
Apologies to Amadeusz since he did almost the same already in transfusion, and I forgot until writing this PR but I also moved the preprocessing to cuda for gpu acceleration.

During tests in my desktop the detection time went from 16.9ms to 13.0ms, making preprocessing about ~1ms if we forcefully add syncs to measure the time.

## Tests performed
 - Tested using data from the taxi

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by knzo25** [comment]
@kminoda 
This is by no means urgent, but it is always nice to reduce the perception latency.

if you do not feel comfortable reviewing this PR, Uetake-san, Hirabayashi-san, Tanaka-san, and Amadeusz are good candidatesI)

**Comment by kminoda** [comment]
@knzo25 Thank you! And yes, I am still catching up with this package (especially the CUDA part), so would be nice if someone else could review it as a co-reviewer :pray: 

@amadeuszsz Hi, would you help me out reviewing CUDA related code?

**Comment by kminoda** [comment]
@knzo25 And also, would you execute Autoware Evaluator and verify that the basic scenarios would pass? Let me know if you don't know how to execute it.

File: perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp
Comment (kminoda) [inline perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp:99]: (let me know if my understanding is wrong :pray: )

My understanding is that this Memcpy from host to device runs each time when the node receives a new pointcloud and hold them as long as the queue size is less than or equal to `param_.pointcloud_cache_size()`.

Here's my question: Is it safe to Memcpy a data, holds it for 0.1[sec] x n seconds, and then use it in `generateSweepPoints`? What would happen if any other process which is also using the same GPU overwrites the copied memory on the GPU device?

**Review by kminoda — COMMENTED** [review]
Thank you for the awesome PR :rocket:

File: perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp
Comment (knzo25) [reply perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp:99]: I am sorry, I do not really understand the question.
If process A allocates N1 bytes in address A1, that address is only known to A.
If processes B,C... only work inside the memory allocates by them, they should never access A1.
cudamalloc of course handles things so that the space [A1, A1 + N1] is reserved by A.

If that was not the question we can talk any day in the office

**Comment by knzo25** [comment]
@kminoda 
Regarding the evaluation I will look into it as soon as I can (got other stuff to deliver first)

**Comment by amadeuszsz** [comment]
> @amadeuszsz Hi, would you help me out reviewing CUDA related code?

@kminoda 
Sure! I would like to make a review next week. Please let me know if it's urgent, I might change my priorities.
I guess you can't add me as a reviewer yet. I added my account to the contributors sheet, so I hope we will be able assign me soon.

File: perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp
Comment (kminoda) [reply perception/lidar_centerpoint/lib/preprocess/pointcloud_densification.cpp:99]: Understood, that answers my question. Thanks!

**Comment by kminoda** [comment]
@amadeuszsz Thank you. Not an urgent task, so please review this when you have time.

File: perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp
Comment (amadeuszsz) [inline perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp:52]:
```
[lidar_centerpoint_node-1]   what():  cudaErrorIllegalAddress (700)@/home/amadeusz/autoware/src/autoware/universe/perception/lidar_centerpoint/include/lidar_centerpoint/cuda_utils.hpp#L68: an illegal memory access was encountered
```
With large input cloud, array may exceed limit. FYI you can observe it if you increase cloud cache size. I think here we can do one of options:
* Check if current `point_counter` + `pc_cache_iter->num_points` is within the limit of `CAPACITY_POINT`. If not, break loop.
* Pass `CAPACITY_POINT` to kernel and monitor current size. If `CAPACITY_POINT` is hit, return.

I'm for the 1st option. With 2nd options we will get one benefit in form of unknown part of cloud, but extra operations in kernel will be necessary.

File: perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp
Comment (amadeuszsz) [reply perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp:63]: I've quickly checked this during the runtime. Number of detections are e.g. (10, 12, 1, 2, 1, 0, 1, 8). After the fix for same input I can see (14, 13, 14, 13, 14, 13, 14, 14). Quality of detections in visualization looks better as well.

File: perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp
Comment (amadeuszsz) [inline perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp:18]: nit: [google style](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes) for headers, especially using `""` instead of `<>` in project's headers. This comment refers to all files of this package.

File: perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp
Comment (amadeuszsz) [inline perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp:67]: `cudaStream_t stream` instead of `cudaStream_t streams` for consistency.

**Review by amadeuszsz — CHANGES_REQUESTED** [review]
Highly performance improvement observed! 👍🏻 
Apart of minor changes requested, please check issues on your desktop with large clouds size & pointer arithmetic which causes wrong detections.

File: perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp
Comment (knzo25) [reply perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp]: Thanks. Addressed in [f3187e9](https://github.com/autowarefoundation/autoware.universe/pull/6989/commits/f3187e9ce8364e86d4539b3ef1fa65223f6a35b9)

Note1: includes is a rabbit hole since even base ros packages are in a chaotic state: https://github.com/ros-perception/image_pipeline/blob/rolling/image_proc/src/rectify.cpp
Note2: The order specified by the google style guidelines can not be used since our pre-commit orders them again

Comment (amadeuszsz) [reply perception/lidar_centerpoint/include/lidar_centerpoint/preprocess/pointcloud_densification.hpp]: Sorry for not precise description: I meant `""` only for `lidar_centerpoint` project as it was shown in this [example](https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/#example).  Anyway, I'm looking into universe project and this pattern is not followed always, therefore fell free to resolve this conversation.

File: perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp
Comment (knzo25) [reply perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp:53]: Thanks for the catch!
Addressed in [80112b1](https://github.com/autowarefoundation/autoware.universe/pull/6989/commits/80112b1be14753f94cde12910e28d1b7fda91366)

File: perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp
Comment (knzo25) [reply perception/lidar_centerpoint/lib/preprocess/voxel_generator.cpp]: Thanks for the catch ! :pray: 
Addressed in [d0ea5f0](https://github.com/autowarefoundation/autoware.universe/pull/6989/commits/d0ea5f0ae60b084d19aa12b11a22d38f2f878569)
(`point_counter` did not mean point counter :melting_face: )

**Comment by knzo25** [comment]
Note to self: Pointpainting was using some of centerpoint parts as a base clase, so I will have to port some of the changes to pointpainting or duplicate logic for pointpainting

**Comment by knzo25** [comment]
@kminoda @YoshiRi @tzhong518 
Since pointainting inherited some logic from centerpoint, and centerpoint now does things differently for GPU acceleration, pointpainting broke. For now, I just copied the logic from centerpoint before this PR and added it to the pointpainting package (the same optimization could not be applied easily).

If you also want pointpainting accelerated, let's us discuss / schedule it a later date please

**Comment by kminoda** [comment]
@knzo25 Thanks. Still haven't understood in depth, but just to confirm: we would anyway have to implement different logic for centerpoint and pointpainting, right? I vote for merging this PR independently from pointpainting's further acceleration :+1:

File: perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/voxel_generator.hpp
Comment (kminoda) [inline perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/voxel_generator.hpp:20]: Do we still need this inclusion?

Comment (knzo25) [reply perception/image_projection_based_fusion/include/image_projection_based_fusion/pointpainting_fusion/voxel_generator.hpp:20]: Yes, the pointpainting one is for the old logic. The centerpoint one is needed since parameter class (for densification) is defined there

**Review by amadeuszsz — APPROVED** [review]

**Comment by amadeuszsz** [comment]
From my side, `lidar_centerpoint` is ready to merge! 🚀 Please, confirm if the `pointpainting` discussion will be solved and we can proceed to merge.

**Review by kminoda — APPROVED** [review]

**Comment by knzo25** [comment]
@kminoda 
I will wait until the message migration happens before attempting to merge this. In addition I want to learn to use the evaluator so I will be trying that as soon as I get a respite from other things :pray:

**Comment by kminoda** [comment]
@knzo25 Hi, would you fix the cppcheck error? (see cppcheck-differential result)

> perception/lidar_centerpoint/include/lidar_centerpoint/network/tensorrt_wrapper.hpp:34:4: warning: inconclusive: Class 'TensorRTWrapper' which has virtual members does not have a virtual destructor. [virtualDestructor]
  ~TensorRTWrapper();

**Comment by xmfcx** [comment]
@knzo25 starting from this PR, The CI for build-main started failing:
- https://github.com/autowarefoundation/autoware.universe/actions/runs/9590002364/job/26444678737#step:15:22008
```
8: C++ exception with description "cudaErrorInsufficientDriver (35)@/__w/autoware.universe/autoware.universe/perception/lidar_centerpoint/include/lidar_centerpoint/cuda_utils.hpp#L80: CUDA driver version is insufficient for CUDA runtime version" thrown in the test body.
```

For some reason the has CI passed for the `build-and-test-differential` cuda yet it fails the `build-and-test` checks.

**Pull request merged**
